shell/platform/darwin/ios/ios_surface_metal.mm

```diff
@@ -3,110 +3,65 @@
-  // various overlays are controlled by this class.
-  return nullptr;
-}
+  auto metal_context = CastToMetalContext(GetContext());
 
-flutter::ExternalViewEmbedder* IOSSurfaceMetal::GetExternalViewEmbedder() {
-  if (IsIosEmbeddedViewsPreviewEnabled()) {
-    return this;
-  } else {
-    return nullptr;
-  }
-}
+  layer_.get().device = metal_context->GetDevice().get();
+  layer_.get().presentsWithTransaction = YES;
 
-void IOSSurfaceMetal::CancelFrame() {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  platform_views_controller->CancelFrame();
-  // Committing the current transaction as |BeginFrame| will create a nested
-  // CATransaction otherwise.
-  [CATransaction commit];
+  is_valid_ = true;
 }
 
-void IOSSurfaceMetal::BeginFrame(SkISize frame_size,
-                                 GrContext* context,
-                                 double device_pixel_ratio) {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  platform_views_controller->SetFrameSize(frame_size);
-  [CATransaction begin];
-}
-
-void IOSSurfaceMetal::PrerollCompositeEmbeddedView(
-    int view_id,
-    std::unique_ptr<flutter::EmbeddedViewParams> params) {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  platform_views_controller->PrerollCompositeEmbeddedView(view_id, std::move(params));
-}
+// |IOSSurface|
+IOSSurfaceMetal::~IOSSurfaceMetal() = default;
 
-// |ExternalViewEmbedder|
-PostPrerollResult IOSSurfaceMetal::PostPrerollAction(
-    fml::RefPtr<fml::GpuThreadMerger> gpu_thread_merger) {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  return platform_views_controller->PostPrerollAction(gpu_thread_merger);
+// |IOSSurface|
+bool IOSSurfaceMetal::IsValid() const {
+  return is_valid_;
 }
 
-std::vector<SkCanvas*> IOSSurfaceMetal::GetCurrentCanvases() {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  return platform_views_controller->GetCurrentCanvases();
+// |IOSSurface|
+void IOSSurfaceMetal::UpdateStorageSizeIfNecessary() {
+  // Nothing to do.
 }
 
-SkCanvas* IOSSurfaceMetal::CompositeEmbeddedView(int view_id) {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  return platform_views_controller->CompositeEmbeddedView(view_id);
+// |IOSSurface|
+std::unique_ptr<Surface> IOSSurfaceMetal::CreateGPUSurface(GrContext* /* unused */) {
+  auto metal_context = CastToMetalContext(GetContext());
+
+  return std::make_unique<GPUSurfaceMetal>(this,    // Metal surface delegate
```

Comment:
Passing a raw c++ pointer inside of a unique_ptr.  Can we use weak_ptr?

Reply:
This is an existing interface that I didn't want to change.